Bug 1698800 - Conform datetime metric reference page to new format (#1611)
mozilla/glean: docs/user/reference/metrics/datetime.md

```

</div>

<div data-lang="Rust" class="tab">

```Rust
use glean_metrics;

use chrono::{FixedOffset, TimeZone};

install::first_run.set(None); // Records "now"
let custom_date = FixedOffset::east(0).ymd(2019, 3, 25).and_hms(0, 0, 0);
install::first_run.set(Some(custom_date)); // Records a custom datetime
```

</div>

<div data-lang="Javascript" class="tab">

```js
import * as install from "./path/to/generated/files/install.js";

install.firstRun.set(); // Records "now"
install.firstRun.set(new Date("March 25, 2019 00:00:00")); // Records a custom datetime
```
</div>

<div data-lang="Firefox Desktop" class="tab">

**C++**

```c++
#include "mozilla/glean/GleanMetrics.h"

PRExplodedTime date = {0, 35, 10, 12, 6, 10, 2020, 0, 0, {5 * 60 * 60, 0}};
mozilla::glean::install::first_run.Set(&date);
```

**Javascript**

```js
const value = new Date("2020-06-11T12:00:00");
Glean.install.firstRun.set(value.getTime() * 1000);
```

</div>

{{#include ../../../shared/tab_footer.md}}

#### Recorded errors

* [`invalid_value`](../../user/metrics/error-reporting.md): Setting the date time to an invalid value.

## Testing API

### `testGetValue`

> Gets the recorded value for a given datetime metric as a language specific Date object.

Reviewer: ```suggestion
Get the recorded value for a given datetime metric as a language-specific Date object.
```

                      year: 2004, month: 12, day: 9, hour: 8, minute: 3, second: 29
                  )
XCTAssertEqual(expectedDate.date!, try Install.firstRun.testGetValue())
```
</div>

<div data-lang="Python" class="tab"></div>

<div data-lang="Rust" class="tab">

```Rust
use glean_metrics;

use chrono::{FixedOffset, TimeZone};

let expected_date = FixedOffset::east(0).ymd(2019, 3, 25).and_hms(0, 0, 0);
assert_eq!(expected_date, metrics.install.first_run.test_get_value(None));
```

</div>

<div data-lang="Javascript" class="tab">

```js
import * as install from "./path/to/generated/files/install.js";

const expectedDate = new Date("March 25, 2019 00:00:00");
assert.deepStrictEqual(expectedDate, await install.firstRun.testGetValue());
```
</div>

<div data-lang="Firefox Desktop" class="tab">

**C++**

```c++
#include "mozilla/glean/GleanMetrics.h"

ASSERT_STREQ(
  mozilla::glean::install::first_run.TestGetValue().value(),
  "2020-11-06T12:10:35+05:00"_ns
);
```

**Javascript**

```js
Assert.ok(Glean.install.firstRun.testGetValue().startsWith("2020-06-11T12:00:00"));
```

</div>
{{#include ../../../shared/tab_footer.md}}

### `testGetValueAsString`

> Gets the recorded value for a given datetime metric an [ISO Date String](https://en.wikipedia.org/wiki/ISO_8601#Dates).

Reviewer: ```suggestion
Get the recorded value for a given datetime metric as an [ISO Date String](https://en.wikipedia.org/wiki/ISO_8601#Dates).
```

assert.strictEqual("2019-03-25-05:00", await install.firstRun.testGetValueAsString());
```
</div>

<div data-lang="Firefox Desktop" class="tab"></div>

{{#include ../../../shared/tab_footer.md}}

### `testHasValue`

Whether or not **any** value was recorded for a given datetime metric.

{{#include ../../../shared/tab_header.md}}

<div data-lang="Kotlin" class="tab">

```Kotlin
import org.mozilla.yourApplication.GleanMetrics.Install

assertTrue(Install.firstRun.testHasValue())
```

</div>

<div data-lang="Java" class="tab">

```Java
import org.mozilla.yourApplication.GleanMetrics.Install;

assertTrue(Install.INSTANCE.firstRun.testHasValue());
```

</div>

<div data-lang="Swift" class="tab">

```Swift
@testable import Glean

XCTAssert(Install.firstRun.testHasValue())
```
</div>

<div data-lang="Python" class="tab">

```Python
from glean import load_metrics
metrics = load_metrics("metrics.yaml")

assert metrics.install.first_run.test_has_value()
```

</div>

<div data-lang="Rust" class="tab"></div>

<div data-lang="Javascript" class="tab"></div>

<div data-lang="Firefox Desktop" class="tab"></div>

{{#include ../../../shared/tab_footer.md}}

### `testGetNumRecordedErrors`

> Gets number of errors recorded for a given counter metric.

Reviewer: ```suggestion
Get number of errors recorded for a given datetime metric.
```

{{#include ../../../shared/tab_header.md}}

<div data-lang="Kotlin" class="tab">

```Kotlin

> import org.mozilla.yourApplication.GleanMetrics.Install

Reviewer: ```suggestion
import mozilla.telemetry.glean.testing.ErrorType
import org.mozilla.yourApplication.GleanMetrics.Install
```

assertEquals(0, Install.firstRun.testGetNumRecordedErrors(ErrorType.InvalidValue))
```

</div>

<div data-lang="Java" class="tab">

```Java

> import org.mozilla.yourApplication.GleanMetrics.Install;

Reviewer: ```suggestion
import mozilla.telemetry.glean.testing.ErrorType
import org.mozilla.yourApplication.GleanMetrics.Install;
```
